### PR TITLE
Add container.cpu.usage metric

### PR DESCRIPTION
**Pull request by ChrsMark**
Related to https://github.com/open-telemetry/semantic-conventions/issues/1032

## Changes

This PR adds the `container.cpu.usage` metric.
This metric is implemented by the [`kubeletstats`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.102.0/receiver/kubeletstatsreceiver/documentation.md#containercpuusage) receiver.
This value comes from the `stats/summary` endpoint of Kubelet and its calculated at https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/kubelet/stats/cri_stats_provider.go#L791-L800.

It's not implemented by the [`dockerstats`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/documentation.md) receiver but that would be doable with something similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.102.0/receiver/dockerstatsreceiver/metric_helper.go#L30, using the formula from the Kubelet's stats mentioned above.

ref: https://docs.docker.com/engine/api/v1.45/#tag/Container/operation/ContainerStats

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

/cc @open-telemetry/semconv-container-approvers @open-telemetry/semconv-k8s-approvers

### CODE REVIEW

**Review by trisch-me — APPROVED** [review]

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:17]: what would be the reason to record it as a `container.cpu.usage` and not one of the existing metrics in `system.cpu.*`? 

I understand there are other metrics in `container` namespace already, and it's be nice to clarify whether we need different ones and if we do, then document why.

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:25]: is this right? should it be time? ticks? percentage?

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:23]: is it normalized per CPU or sum across all of them?

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:21]: could we clarify a range of this value in the note?

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:28]: > A container SHOULD be characterized _either_ by data points with no `state` labels, _or only_ data points with `state` labels."

Could it be clarified?
cpu state does not characterizes a container, but a state of the CPU. Also, does it mean that cpu.state should be added on all data points or not applied at all for a given container?

File: docs/system/container-metrics.md
Comment (lmolkova) [reply docs/system/container-metrics.md:98]: I think `cpu.mode` would be more accurate, but we can follow up in the https://github.com/open-telemetry/semantic-conventions/issues/1139

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:23]: [kubeletstats](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.102.0/receiver/kubeletstatsreceiver/documentation.md#containercpuusage) reports the total on all available CPUs, does not break down per cpu.
[dockerstats](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/documentation.md) does the same but maybe in the future it can become more granular since [dockerAPI](https://docs.docker.com/engine/api/v1.45/#tag/Container/operation/ContainerStats) can provide per cpu metrics. In any case, if we'd like to change it the future we can add the `cpu.logical_number` attribute as `opt-in` similarly to [`cpu.state`](https://github.com/open-telemetry/semantic-conventions/pull/1128#discussion_r1634449851)? (that was also discussed at https://github.com/open-telemetry/semantic-conventions/pull/282#issuecomment-1909118310).

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:25]: Based on the implementation from [kubeletstats](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.102.0/receiver/kubeletstatsreceiver/documentation.md#containercpuusage) it is `{cpu}`, see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25901#issuecomment-1885979641 for more context.

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:17]: It is documented at https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/system/system-metrics.md?plain=1#L15-L17.

Comment (lmolkova) [reply model/metrics/container.yaml:17]: thanks! would you mind adding a similar statement to container metrics (that they are not the same as system ones)?

do you know why - would system metrics be different than container metrics? or it's a way to distinguish them if both sets of metrics are reported?

File: docs/system/container-metrics.md
Comment (TylerHelmuth) [reply docs/system/container-metrics.md:98]: Is there a conclusion from https://github.com/open-telemetry/semantic-conventions/issues/840 that we'll stick with `container.cpu.state`?

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:17]: Mostly a way to distinguish them since a container (or Pod) is not a "system"(=="host") neither a standalone process (it's a group of processes, aka cgroup). I wouldn't be surprised if some API specific metrics might differ and extra transformations would be needed to make them match the `system.*` metrics, so that would be an extra pain/hassle for the implementations I guess. Also it depends from which point of view we observe a container. If we run a collector inside a container treating it as a host then from this point of view the container is a "host" and hence `system.*` metrics should be collected using the `hostmetrics` receiver. When we observe containers/pods from outside, using a centralized API, we treat them mostly as "processes" (but strictly speaking a container can include more than 1 process). Since those entities are technology specific, we decided back then that for centralized metrics that are collected from outside of the resource's scope (ie container metrics collected centrally for a k8s cluster) we use resource specific namespace like `container.*` or `k8s.pod`.
I think this is somehow similar to the `jvm.*` namespace.

Some relevant discussions, issues:
- https://github.com/open-telemetry/semantic-conventions/issues/226#issuecomment-1700529192
- https://github.com/open-telemetry/opentelemetry-specification/pull/2388#issuecomment-1057120767

Comment (lmolkova) [reply model/metrics/container.yaml:17]: thank you for the context!

File: docs/system/container-metrics.md
Comment (heyams) [inline docs/system/container-metrics.md:13]:
```suggestion
metrics in OpenTelemetry. These metrics are collected from technology-specific,
well-defined APIs (e.g. Kubelet's API or container runtimes).
```

File: model/metrics/container.yaml
Comment (lmolkova) [reply model/metrics/container.yaml:25]: In all other utilization metrics we use `1` as a unit. Having `{cpu}` there creates an impression that we measure number of CPUs.

File: model/metrics/container.yaml
Comment (lmolkova) [reply model/metrics/container.yaml:23]: So if I have 10 CPUs that are busy 100% of time, would this metric show 100% (1) or 1000% (10) ? Could you please clarify the brief so it's clear to the future readers. See my other comment on the value range as well.

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:28]: maybe (to make it a bit more clear, also feel free to move it back to brief if you like it better)
```suggestion
        brief: "The CPU state for this data point."
        note: >
          `container.cpu.state` SHOULD be reported consistently on all measurements if it's reported at all.
```

File: model/metrics/container.yaml
Comment (lmolkova) [inline model/metrics/container.yaml:29]: Why not `recommended`? Having one dimension with 3 possible values on the metric should be fine.

File: model/metrics/container.yaml
Comment (lmolkova) [reply model/metrics/container.yaml:23]: In other places we call a similar metric `*.cpu.utilization`  and measure the normalized one (divided by cpu count) and in a [0, 1] range.  (see JVM, process, system metrics)

It'd be inconsistent if we had a different name and measured a sum across all CPUs here. 

We also report `*.cpu.count` in other places (as a metric, it has high cardinality and can't be used as an attribute).
The `cpu.time` and `cpu.count` metrics can be used together to calculate utilization. Utilization is still reported as a relatively cheap convenience.

By introducing non-normalized `cpu.usage` in addition to `cpu.time` here we provide no way for users to calculate normalized thing and create any alerts.

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:29]: As explained also at https://github.com/open-telemetry/semantic-conventions/pull/1128#discussion_r1634449851, `kubeletstats` cannot distinguish the state/mode while the `dockerstats` can. That's why we marked it as opt-in when we [introduced that](https://github.com/open-telemetry/semantic-conventions/pull/282).

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:28]: I'm fine with changing this, however we define it like that in [system](https://github.com/open-telemetry/semantic-conventions/blob/main/model/metrics/system-metrics.yaml#L12) and [process](https://github.com/open-telemetry/semantic-conventions/blob/main/model/metrics/process-metrics.yaml#L12) metrics. Do you think diverging them would be ok?

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:25]: But this is not a utilization metric, this is usage metric.
Correct me if I'm wrong here but the intention is to measure the number of CPUs used by the container over a reporting interval. That means essentially that if a container uses 2 complete CPUs constantly during the internal, the metric would be `2`. 
Then if we divide this with the number of Node's cores for example (ie `4` cores) it would give us `2/4=0.5` utilization.
In `kubeletstats` we have other metrics specifically for reporting the utilization against the container's limits and/or the Node's capacity.

cc @TylerHelmuth @dmitryax @povilasv who were involved in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25901.

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml:23]: > So, the suggestion:
>
> add container.cpu.logical_count or similar metric
> add container.cpu.utilization metric (in [0,1] range) instead of container.cpu.usage

It has been preferred to use the `container.cpu.usage` in `kubeletstats` receiver after a change at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25901.

As I also mention at https://github.com/open-telemetry/semantic-conventions/pull/1128#discussion_r1642699619 the `*utilization` metrics can be calculated in various different ways in K8s against various different limits. For sure we are planning to add those `*utilization` metrics as SemConvs as part of the https://github.com/open-telemetry/semantic-conventions/issues/1032. 
So I guess that `container.cpu.usage` can be added independently of the `*utilization` ones.

File: docs/system/container-metrics.md
Comment (ChrsMark) [reply docs/system/container-metrics.md]: Switched to `cpu.mode` since https://github.com/open-telemetry/semantic-conventions/issues/840 was merged. We can resolve this thread I guess.

File: model/metrics/container.yaml
Comment (dmitryax) [reply model/metrics/container.yaml:26]: Correct. We use `utilization` when there is a limit available and count it as a percentage [0,1]. Aggregation of those metrics would be `avg` which result in the same units `1`.

Here, we have an average usage of CPUs. If we had `cpu.logical_number` attribute from docker, the aggregation method would be `sum`, which gives us the same `{cpu}` unit.

**Review by dmitryax — APPROVED** [review]

**Comment by ChrsMark** [comment]
@open-telemetry/semconv-k8s-approvers @open-telemetry/semconv-container-approvers @open-telemetry/specs-semconv-approvers could you take a look?

File: model/metrics/container.yaml
Comment (jsuereth) [reply model/metrics/container.yaml:31]: You could also use conditionally required (where condition is dockerstats)

File: model/metrics/container.yaml
Comment (mx-psi) [inline model/metrics/container.yaml:22]:
```suggestion
    brief: "Container's CPU usage, measured in cpus. Range from 0 to the number of allocatable CPUs"
```

**Review by mx-psi — APPROVED** [review]

File: model/metrics/container.yaml
Comment (ChrsMark) [reply model/metrics/container.yaml]: Thank's for the suggestion. Added at [9bc986b](https://github.com/open-telemetry/semantic-conventions/pull/1128/commits/9bc986b361db6144c768d53590fed1ac86a757f3). For some reason it has no impact on the MD table though 🤔 (even after running `make fix` or `make table-generation `).

**Comment by ChrsMark** [comment]
@open-telemetry/specs-semconv-approvers since there are quite a few approvals by the k8s/container approvers can we move this one forward?

**Review by joaopgrassi — APPROVED** [review]

**Comment by ChrsMark** [comment]
Thank's @joaopgrassi! My only concern is about https://github.com/open-telemetry/semantic-conventions/pull/1128#discussion_r1686534350. Does this look normal?

File: model/metrics/container.yaml
Comment (joaopgrassi) [reply model/metrics/container.yaml]: We spoke on slack, @lquerel will investigate what looks like a bug in the tooling.

Comment (ChrsMark) [reply model/metrics/container.yaml]: I think we can move this forward by changing the requirement level in the `container.cpu.time` as well, which is something needed anyways. And the issue in the tooling can be investigated independently.

File: docs/system/container-metrics.md
Comment (joaopgrassi) [inline docs/system/container-metrics.md:102]: This is still wrong, as @lquerel spotted. @ChrsMark curious now, did you add this manually or was it produced by the markdown generation? We aren't supposed to edit these markdown files manually anymore so maybe that was it? 🤔

Comment (ChrsMark) [reply docs/system/container-metrics.md:102]: Good catch, I didn't notice that line. It was automatically generated by the tooling. I just changed the `requirement_level` of the `container.cpu.time` metric.
I see, we need to block this until the tooling is fixed.

Comment (joaopgrassi) [reply docs/system/container-metrics.md:102]: Credits to @lquerel good eyes 😅

Comment (lmolkova) [reply docs/system/container-metrics.md:102]: wait, why are we not supposed to edit these files? It's not the registry, so we added `container.cpu.time` and the fix would be

```suggestion
<!-- semconv metric.container.cpu.usage(full) -->
```

what am I missing?

File: model/metrics/container.yaml
Comment (lmolkova) [reply model/metrics/container.yaml]: As I mentioned in the other comment, there seem to be no issue on tooling side, we just need to use proper group id (related to cpu.usage and not time)

**Review by lmolkova — APPROVED** [review]
Please fix anchor name and this PR should be good to go.

File: docs/system/container-metrics.md
Comment (lmolkova) [reply docs/system/container-metrics.md]: no worries! It seems everything is fixed now.

**Pull request merged**
